### PR TITLE
Cache policy control for image source

### PR DESCRIPTION
**Pull request by salanki**
## Motivation
In the context of an app an image exists in three resolutions on the server: `thumb` (30px) `feed` (300px) `full` (900px). When looking at an individual item a user can come either from the feed, via a permalink or from other parts of the app. This allows a situation where the `feed` image might or might not already be loaded somewhere in the app. In the detail view I want to render `thumb` with a blur (to quickly display something), then the `feed` image if it exists to have something decent to display until `full` loads. However it is quite a waste to load the `feed` image if it isn't already in cache, and will slow down the time until `full` is loaded. It is possible to track the navigation from feed->detail and that the `feed` image has actually completed loading by the feed component however as component hierarchies grow this turns into quite a lot of prop passing and bad separation of concerns.

## Solution
NSURLRequests accepts a [Cache Policy](https://developer.apple.com/reference/foundation/nsurlrequest.cachepolicy) that can be used to control a requests interaction with the cache. For the use case in motivation a cache policy of `returnCacheDataDontLoad` fulfills the requirement.

This PR exposes the cace policy to the Image Source. Cache policy is controlled via a `cache` property. The [fetch](https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch) API provides a platform agnostic way to specify cache policy. In an effort to provide an API ready for a future Android implementation, and users with something they recognize from the web the property naming and values are identical to the ones in `fetch`. The values are mapped against the iOS API per the table below.

|iOS|	fetch (and this PR)|
|----|-----|
|useProtocolCachePolicy|default|
|reloadIgnoringLocalCacheData|reload|
|returnCacheDataElseLoad|force-cache|
|returnCacheDataDontLoad|only-if-cached|

## Notes

* An example is added to UIExplorer for both demonstration and testing purposes, however this can be removed if deemed redundant.

## Test Plan

An example has been added to UIExplorer for both demonstration and testing purposes. This example demonstrates the two cases (`default` and `only-if-cached`) that can be proven without a custom server component. As all we are doing is setting different native properties it would be safe to assume that if two can be set the others will be set correctly as well (or there is a serious bug in `RCT_ENUM_CONVERTER`)

### CODE REVIEW

File: Libraries/Image/ImageSourcePropType.js [hunk 1]
```diff
@@ -43,6 +43,20 @@ const ImageURISourcePropType = PropTypes.shape({
    */
   body: PropTypes.string,
   /**
+   * `cachePolicy` determines how the requests handles potentially cached
+   * responses. These directly match the underlying iOS policies.
+   * Default is; `useProtocolCachePolicy`
+   * https://developer.apple.com/reference/foundation/nsurlrequestcachepolicy
+   */
```
Comment (leeight) [inline Libraries/Image/ImageSourcePropType.js]: `@platform ios` ??

Comment (salanki) [inline Libraries/Image/ImageSourcePropType.js]: Fixed!

**Comment by oblador** [comment]
This is great! As a general feedback, most enum prop values throughout the codebase use kebab-case. Before you change that though I would like to discuss the enum itself as it would probably be a good idea to think of future android implementation. I'm coming up with two ways to go:

* Don't use platform specific wording, but borrow web semantics instead. `Cache-Control` header comes to mind.
* Use a suffix, ie `cachePolicyIOS` for now and when we have a cross platform API we deprecate it.

**Comment by salanki** [comment]
A platform agnostic property would be better than a iOS specific one. I took a stab at mapping these to the HTTP Cache-Control headers. Only a few of them are good matches, I came up with a home-made terminology in the alternative column that better matches what it actually does. Let me know what you think.

| iOS                                   | Cache-Control   | Good Match                                                 | Alternative        |
|---------------------------------------|-----------------|------------------------------------------------------------|--------------------|
| useProtocolCachePolicy                |                 | No. There is no match.                                     | default            |
| reloadIgnoringLocalCacheData          | no-store        | Semi-good                                                  | ignore-local-cache |
| reloadIgnoringLocalAndRemoteCacheData | no-cache        | So-so                                                      | ignore-all-cache   |
| returnCacheDataElseLoad               | max-age         | Not great, people are used to providing a value to max-age | if-exists          |
| returnCacheDataDontLoad               | only-if-cached  | Great                                                      | only-if-cached     |
| reloadRevalidatingCacheData           | must-revalidate | Semi-good                                                  | must-revalidate    |

**Comment by oblador** [comment]
`reloadIgnoringLocalAndRemoteCacheData` and `reloadRevalidatingCacheData` [aren't implemented on iOS](http://openradar.appspot.com/radar?id=1755401) so we can ignore those for now. There's another API we can borrow from and that's `fetch` (although it's not implemented in the what-wg polyfill). [MDN](https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch) says: 
> cache: The cache mode you want to use for the request: default, no-store, reload, no-cache, force-cache, or only-if-cached.

This is my proposed mapping:

iOS|fetch
------------ | -------------
useProtocolCachePolicy|default
reloadIgnoringLocalCacheData|reload
returnCacheDataElseLoad| force-cache
returnCacheDataDontLoad|only-if-cached

**Comment by salanki** [comment]
PR updated based on feedback from @oblador. Stealing the `fetch` API is a great idea to be platform agnostic and to give an API that people can recognize from the web. I also added some more documentation as the JSDoc in the proptype definition never makes it to the generated HTML.

**Comment by oblador** [comment]
LGTM 👍 @hramos & @javache?

File: Libraries/Image/ImageSourcePropType.js [hunk 2]
```diff
@@ -43,6 +43,32 @@ const ImageURISourcePropType = PropTypes.shape({
...
+   * responses.
+   *
+   * - `default`: Use the native platforms default strategy. `useProtocolCachePolicy` on iOS.
+   *
+   * - `reload`: The data for the URL will be loaded from the originating source.
+   * No existing cache data should be used to satisfy a URL load request.
+   *
+   * - `force-cache`: The existing cached data will be used to satisfy the request,
+   * regardless of its age or expiration date. If there is no existing data in the cache
+   * corresponding the request, the data is loaded from the originating source.
```
Comment (mkonicek) [inline Libraries/Image/ImageSourcePropType.js]: Not sure if force-cache is the right name here as this will still make a network request if not present in cache?

File: docs/Images.md [hunk 3]
```diff
@@ -91,6 +91,26 @@ Many of the images you will display in your app will not be available at compile
...
+* `default`: Use the native platforms default strategy.
+* `reload`: The data for the URL will be loaded from the originating source.
+No existing cache data should be used to satisfy a URL load request.
+* `force-cache`: The existing cached data will be used to satisfy the request,
+regardless of its age or expiration date. If there is no existing data in the cache
+corresponding the request, the data is loaded from the originating source.
+* `only-if-cached`: The existing cache data will be used to satisfy a request, regardless of
+its age or expiration date. If there is no existing data in the cache corresponding
+to a URL load request, no attempt is made to load the data from the originating source,
+and the load is considered to have failed.
```
Comment (mkonicek) [inline docs/Images.md]: By duplicating the docs anyone making changes in the future will have to make sure they are updated in both places.

But since it looks like `ImageSourcePropType` is not rendered on the website it makes sense to also add the docs in this guide.

File: React/Base/RCTConvert.m [hunk 4]
```diff
@@ -140,8 +148,9 @@ + (NSURLRequest *)NSURLRequest:(id)json
 
     NSData *body = [self NSData:json[@"body"]];
     NSString *method = [self NSString:json[@"method"]].uppercaseString ?: @"GET";
+    NSURLRequestCachePolicy cachePolicy = [self NSURLRequestCachePolicy:json[@"cache"]];
```
Comment (mkonicek) [inline React/Base/RCTConvert.m]: Does this return `nil` if not present?

**Comment by mkonicek** [comment]
How have you tested all four options work as documented? `ImageExample.js` only shows 2/4 values:

|iOS|	fetch (and this PR)|
|----|-----|
|useProtocolCachePolicy|default|
|reloadIgnoringLocalCacheData|reload|
|returnCacheDataElseLoad|force-cache|
|returnCacheDataDontLoad|only-if-cached|

Please add a Test Plan showing screenshots from the UIExplorer for all 4 values.

In general this looks good and useful, thanks for working on it!

**Comment by mkonicek** [comment]
Requesting changes mainly because of a missing Test Plan, see:
https://github.com/facebook/react-native/blob/master/.github/PULL_REQUEST_TEMPLATE.md

File: Libraries/Image/ImageSourcePropType.js [hunk 5]
```diff
@@ -43,6 +43,32 @@ const ImageURISourcePropType = PropTypes.shape({
    */
   body: PropTypes.string,
   /**
+   * `cache` determines how the requests handles potentially cached
+   * responses.
+   *
+   * - `default`: Use the native platforms default strategy. `useProtocolCachePolicy` on iOS.
+   *
+   * - `reload`: The data for the URL will be loaded from the originating source.
+   * No existing cache data should be used to satisfy a URL load request.
```
Comment (salanki) [inline Libraries/Image/ImageSourcePropType.js]: I do not personally have a use case for it, however I implemented it for completeness since it exists both in iOS and in `fetch`. The only use case I can think of is when the server or intermediate proxies can't be trusted to set a appropriately low cache timeout for volatile content.

File: Libraries/Image/ImageSourcePropType.js (hunk 2 above)
Comment (salanki) [inline Libraries/Image/ImageSourcePropType.js]: This terminology and behavior matches directly with the `fetch` API (please see earlier discussion in the thread). Even though it might be argued that the name is not descriptive enough, I believe there is a far bigger benefit sticking with already accepted and well used terminology than inventing our own.

File: Libraries/Image/ImageSourcePropType.js [hunk 6]
```diff
@@ -43,6 +43,32 @@ const ImageURISourcePropType = PropTypes.shape({
    */
   body: PropTypes.string,
   /**
+   * `cache` determines how the requests handles potentially cached
+   * responses.
+   *
+   * - `default`: Use the native platforms default strategy. `useProtocolCachePolicy` on iOS.
```
Comment (salanki) [inline Libraries/Image/ImageSourcePropType.js]: The best documentation I can point you to is https://developer.apple.com/reference/foundation/nsurlrequest.cachepolicy. I do not know the exact details of how the iOS caching works by default, but I would assume that it will simply serve it from local cache as long as it is not expired, following HTTP standards. It is safe to assume that it would not act as `force-cache` as that is not the expected behavior according to the protocol.

Basically, this default is how things act when there is no cache policy set, ie. how all requests behave in RN today.

File: docs/Images.md (hunk 3 above)
Comment (salanki) [inline docs/Images.md]: I had to pick between a rock and a hard place. It didn't make sense to only document it in `ImageSourcePropType` as that would never be visible in the docs. If I only documented it in `Images` or the `Image` component I would be breaking the pattern already established inside `ImageSourcePropType` where there is documentation for all properties (even though it's not visible anywhere outside of the source).

**Comment by salanki** [comment]
@mkonicek: Thank you for your review. I addressed your questions in comments. When it comes to the test plan I figured that if I could show the general concept working that should be enough, since all I am doing is choosing different native properties. Adding test cases for all 4 options would really just test the Apple NSURLSession code and not add any additional integrity to this PR.

The reason I am pushing back on testing of the other two methods is that it would require a custom server to actually demonstrate it, and some coordination between the server and client. Just an image hosted somewhere won't do, as the server will need to first send one image on a  URL and then send another for a subsequent request from the same client (all while setting long cache timeout headers). I wouldn't have a problem building a small node app and hosting it myself for this purpose, but I am sure you want to limit the number of corners of the internet that UIExplorer depends on. Please advice on how I can proceed.

File: Libraries/Image/ImageSourcePropType.js (hunk 5 above)
Comment (salanki) [inline Libraries/Image/ImageSourcePropType.js]: @oblador: Do you happen to have a use case?

**Comment by lacker** [comment]
Can this be unit tested rather than only having an example for testing?

**Comment by salanki** [comment]
@lacker: Is there a unit testing framework for images already in place, ie. being able to compare image A to image B? If not it feels like the infrastructure I need to invent to test it will be 10x the actual implementation :/

**Comment by lacker** [comment]
I don't think you would want to actually load up images into pixels and compare them, you should probably stub that out somehow. Isn't there testing for the Image component somewhere, maybe inside the UI explorer?

**Comment by salanki** [comment]
@lacker: Sure, but I would want to compare the byte data to see that the right image is loaded. I have looked everywhere I can think of (including the UIExplorer tests) and can't find any tests for the Image component. If there is and I can add on to that sure, but if not building a test suite for the image component seems like another PR.

**Comment by lacker** [comment]
I dunno, it seems reasonable to add a test as part of this one. How hard can it be? Also you would be an amazing human being, a truly impressive open source machine, bringing high quality code to the world and really pushing humanity forward.

**Comment by salanki** [comment]
Well when you put it that way! I'll get to it!

**Comment by hramos** [comment]
How are those tests going? :)

**Comment by salanki** [comment]
Hello @lacker @hramos. I have been trying to figure out how to integrate a proper test for this. Since it is testing native code it needs to live in `IntegrationTests`. However, to test the `reloadIgnoringLocalCacheData` and `returnCacheDataElseLoad` a special web-server (can be written in node) is required. I see two options in implementing this:

1) Host the node server externally ie. on Heroku, and access via internet. This is not a very appealing option and I'm sure wouldn't get accepted.
2) Change the integration test framework to allow launching a node server. How to do that in an acceptable way I'd need help with, and where in the source repo it should live. Maybe it can be started from `scripts/objc-test.sh` or should it be started in the xcode build steps somehow?

The third option is to remove `reloadIgnoringLocalCacheData` and `returnCacheDataElseLoad` since testing of those two options are such a pain. Let me know.

**Comment by salanki** [comment]
I added the most complete tests I could without spinning up a custom web server. They use `https://facebook.github.io/react/img/logo_small_2x.png` over the internet (got the URL from UIExplorer).

**Comment by salanki** [comment]
@lacker will you be able to review my last additions? Thank you!

**Comment by lacker** [comment]
Looks good - thanks @salanki !

**Comment by salanki** [comment]
Thank you @lacker!!
